notes.md

```diff
@@ -0,0 +1,179 @@
+   - gracefully handle these scenarios itself (auto-recover if possible and makes sense), or
+   - permit the user to handle them gracefully, or 
+   - notify the user of this non-fatal error?
+5) (Visibility) ...
+6) (Efficiency + Performance) ...
+
+## Use Cases
+Below are the use-cases FS2-Netty needs to support.
+
+### A Pub-Sub like System
+This system uses the websocket protocol to integrate with clients. On the backend, messages can be sent via HTTP/1.1.
+Published events from clients are arbitrarily processed. The following are requirements of a networking framework from
+the point of view of developers building the pub-sub system. 
+
+#### High Level Requirements
+HTTP:
+- We want a simple (GET & POST) but robust (handles malformed requests) HTTP server that handles payloads on the order of KB's.
+- We want HTTP connections to be efficient for client use, i.e. a client can pool connections and reuse them for multiple HTTP requests.
+- We want the HTTP server to avoid leaking connections; it should detect and clean up dropped connections.
+- We want the HTTP server to rotate client connections to load balance across multiple instances.
+- We don't want to allow HTTP pipelining.
+- We want HTTP logging (and metrics?).
+  - Access log with URI path, HTTP method, HTTP headers, payload size in bytes, response time, response code, and response headers.
+  - Metrics on how often the TCP connection is closed by the server and the client. How often client drops connections? 
+- We don't need SSL since this service will be behind a Load Balancer that will terminate SSL connections.
+  
+WebSocket (at a minimum): 
+- We want the HTTP connection to transition to an arbitrary WebSocket connection, i.e. there's not a single WS controller with a single path for the whole server.
+- We want to receive websocket frames.
+  - We want to backpressure the sender. For context, say our service proxies frames to other tcp servers. Furthermore,
+    one of those servers is slow. We want the client to slow down their rate without an explicit higher layer protocol
+    message. However, we can optimize this so that a single slow downstream does backpressure all other frames going to
+    other destinations. We maintain a queue of frames and only backpressure when queue is full. (Without an explicit
+    higher layer flow control or protocol other TCP, this will only mitigate the problem. Frames for different
+    downstreams are multiplexed over a single TCP connection, so once queue is full because a single slow consumer than 
+    we block all other frame processing, a.k.a Head-of-line (HOL) blocking. This is partly a fundamental limitation of
+    TCP, i.e. there's no virtual streams. Such limitations are overcome in QUIC and HTTP/3.)
+    Either of these 2 approaches may work:
+    - Only read socket when we finish processing a message. Pub-Sub app may queue up the frame for later processing, or it may immedieiately process it.
```

Comment:
Yeah, going to presume this will be the way going forward. Curious if we'll need to interact with `channelReadComplete` for pipelines that automatically aggregate, e.g. an HTTP server with `HttpObjectAggregator` in the pipeline. Basically, might have to call read until `channelReadComplete` is hit. I never messed with this method. Will write tests around it, but think it's worth calling out.